### PR TITLE
New setting for maintenance mode

### PR DESCRIPTION
**Pull request by yurabakhtin**
**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [x] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [ ] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] It's submitted to the `develop` branch, _not_ the `master` branch if no hotfix
- [ ] When resolving a specific issue, it's referenced in the PR's description (e.g. `Fix #xxx[,#xxx]`, where "xxx" is the Github issue number)
- [ ] All tests are passing
- [ ] New/updated tests are included
- [x] Changelog was modified

If adding a **new feature**, the PR's description includes:
- [ ] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

**Other information:**
Issue: https://github.com/humhub/humhub/issues/3345

### CODE REVIEW

File: protected/humhub/modules/user/controllers/AuthController.php [hunk 1]
```diff
@@ -103,6 +103,10 @@ public function actionLogin()
             }
         }
 
+        if (Yii::$app->settings->get('maintenanceMode')) {
+            Yii::$app->session->setFlash('error', Yii::t('error', 'Only admins have an access to the site on maintenance mode.'));
```
Comment (ArchBlood) [inline protected/humhub/modules/user/controllers/AuthController.php:107]: You can remove `an` it's not needed

File: protected/humhub/modules/user/models/Invite.php [hunk 2]
```diff
@@ -112,6 +112,11 @@ public function beforeSave($insert)
 
     public function selfInvite()
     {
+        if (Yii::$app->settings->get('maintenanceMode')) {
+            Yii::$app->getView()->warn(Yii::t('error', 'Only admins have an access to the site on maintenance mode.'));
```
Comment (ArchBlood) [inline protected/humhub/modules/user/models/Invite.php:116]: Here as well `an` can be removed

File: protected/humhub/components/behaviors/AccessControl.php [hunk 3]
```diff
@@ -237,4 +237,19 @@ protected function forceChangePassword()
...
+     * Log out all non admin users when maintenance mode is active
+     *
+     * @return Response Redirect to home page
+     * @since 1.8
+     */
+    protected function checkMaintenanceMode()
+    {
+        if (Yii::$app->settings->get('maintenanceMode') && !Yii::$app->user->isGuest) {
+            Yii::$app->user->logout();
+            Yii::$app->getView()->warn(Yii::t('error', 'You have been logged out automatically because only admins have an access to the site on maintenance mode!'));
```
Comment (ArchBlood) [inline protected/humhub/components/behaviors/AccessControl.php:251]: `an` isn't needed

File: protected/humhub/components/access/ControllerAccess.php [hunk 4]
```diff
@@ -244,6 +244,12 @@ public function init()
             'reason' => Yii::t('error', 'Login required for this section.'),
             'code' => 401
         ]);
+        $this->registerValidator([
+            self::RULE_LOGGED_IN_ONLY => 'validateMaintenanceMode',
+            'reason' => Yii::t('error', 'Only admins have an access to the site on maintenance mode.'),
```
Comment (ArchBlood) [inline protected/humhub/components/access/ControllerAccess.php:249]: Here as well `an` not really needed

**Review by ArchBlood — COMMENTED** [review]
Here's a few things when it comes to the messages used, `an` isn't needed in these instances and can be removed :+1:

File: protected/humhub/modules/admin/widgets/views/maintenanceModeWarning.php [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+use humhub\libs\Html;
+use yii\helpers\Url;
+
+?>
+
+<div class="panel panel-danger panel-invalid">
+    <div class="panel-heading"><?= Yii::t('AdminModule.base', '<strong>Maintenance</strong> mode!'); ?></div>
+    <div class="panel-body">
+        <p><?= Yii::t('AdminModule.base', 'Maintenance Mode is Active and only administrators have access.'); ?></p>
```
[the remark this answers is not in the document]
Comment (yurabakhtin) [reply protected/humhub/modules/admin/widgets/views/maintenanceModeWarning.php]: Commit https://github.com/humhub/humhub/pull/4773/commits/23a9ed3c0a5750e569d9480669fc7fe0d6f52ae1

File: protected/humhub/modules/user/controllers/AuthController.php (hunk 1 above)
Comment (yurabakhtin) [reply protected/humhub/modules/user/controllers/AuthController.php]: Commit https://github.com/humhub/humhub/pull/4773/commits/23a9ed3c0a5750e569d9480669fc7fe0d6f52ae1

File: protected/humhub/modules/user/models/Invite.php (hunk 2 above)
Comment (yurabakhtin) [reply protected/humhub/modules/user/models/Invite.php]: Commit https://github.com/humhub/humhub/pull/4773/commits/23a9ed3c0a5750e569d9480669fc7fe0d6f52ae1

File: protected/humhub/components/behaviors/AccessControl.php (hunk 3 above)
Comment (yurabakhtin) [reply protected/humhub/components/behaviors/AccessControl.php]: Commit https://github.com/humhub/humhub/pull/4773/commits/23a9ed3c0a5750e569d9480669fc7fe0d6f52ae1

File: protected/humhub/components/access/ControllerAccess.php (hunk 4 above)
Comment (yurabakhtin) [reply protected/humhub/components/access/ControllerAccess.php]: Commit https://github.com/humhub/humhub/pull/4773/commits/23a9ed3c0a5750e569d9480669fc7fe0d6f52ae1

File: protected/humhub/modules/user/views/auth/login_modal.php [hunk 6]
```diff
@@ -34,14 +34,21 @@
             <div class="tab-content">
                 <div class="tab-pane <?= (!isset($_POST['Invite'])) ? "active" : ""; ?>" id="login">
 
+                    <?php if (Yii::$app->session->hasFlash('error')): ?>
```
Comment (ArchBlood) [inline protected/humhub/modules/user/views/auth/login_modal.php:37]: In this case, would `error` being used over `warning` be necessary? Using error seems a little extreme in this case

**Review by ArchBlood — COMMENTED** [review]
Just a quick thought, can be looked over if need be

File: protected/humhub/modules/user/views/auth/login_modal.php (hunk 6 above)
Comment (luke-) [reply protected/humhub/modules/user/views/auth/login_modal.php:37]: In the non-modal login view there is currently only an "Error" output. Adding a warning block in both Views would be too much and we try to keep View changes as little as possible (or at least to combine them with other changes).

**Review by luke- — CHANGES_REQUESTED** [review]
Looks fine!

- Seems the "admin.changeMaintenanceMode" javascript handle is not included in the PR.
- It would be good if we had a small text field for a free text (single-line input) which is also displayed on the login page. An admin could enter the duration here, for example.

**Comment by yurabakhtin** [comment]
@luke- 

> Seems the "admin.changeMaintenanceMode" javascript handle is not included in the PR.

It is included here:

- https://github.com/humhub/humhub/pull/4773/files#diff-291adf6dd0e6067fb59ee9396f7a8460d01ebbc10ff011631cad5df35cf69c12R126
- https://github.com/humhub/humhub/pull/4773/files#diff-291adf6dd0e6067fb59ee9396f7a8460d01ebbc10ff011631cad5df35cf69c12R140

Probably you should update the cached file `assets/9047f518/js/humhub.admin.js`, or force copy in the file `protected/humhub/modules/admin/assets/AdminAsset.php`.

**Comment by yurabakhtin** [comment]
@luke- 

> Seems the "admin.changeMaintenanceMode" javascript handle is not included in the PR.

It is included here:

- https://github.com/humhub/humhub/pull/4773/files#diff-291adf6dd0e6067fb59ee9396f7a8460d01ebbc10ff011631cad5df35cf69c12R126
- https://github.com/humhub/humhub/pull/4773/files#diff-291adf6dd0e6067fb59ee9396f7a8460d01ebbc10ff011631cad5df35cf69c12R140

Probably you should update the cached file `assets/9047f518/js/humhub.admin.js`, or force copy in the file `protected/humhub/modules/admin/assets/AdminAsset.php`.

**Comment by luke-** [comment]
@yurabakhtin I tried to flush the assets without success. See here:
https://github.com/humhub/humhub/pull/4773/files#diff-b979536d7e8dde5a63337736cbf1a17b5f0ec9ed1ae32972402f90be2c73f4a0R63
image

**Comment by yurabakhtin** [comment]
@luke- Please try to find the file `assets/9047f518/js/humhub.admin.js`, probably you have another value  instead of `9047f518`, because source file is in `protected/humhub/modules/admin/resources/js/humhub.admin.js`.
I think you should remove the file `assets/9047f518/js/humhub.admin.js` manually and refresh page in order to update it from the source.

**Comment by yurabakhtin** [comment]
@luke- 

> It would be good if we had a small text field for a free text (single-line input) which is also displayed on the login page. An admin could enter the duration here, for example.

Commit https://github.com/humhub/humhub/pull/4773/commits/5ba6603f553573dec1054e0b2c822c1263661ef5:
mode_info

File: protected/humhub/components/behaviors/AccessControl.php [hunk 7]
```diff
@@ -248,7 +248,7 @@ protected function checkMaintenanceMode()
     {
         if (Yii::$app->settings->get('maintenanceMode') && !Yii::$app->user->isGuest) {
             Yii::$app->user->logout();
-            Yii::$app->getView()->warn(Yii::t('error', 'You have been logged out automatically because only admins have access to the site on maintenance mode!'));
+            Yii::$app->getView()->warn(Yii::t('error', 'Maintenance mode activated: You have been automatically logged out and will no longer have access the platform until the maintenance has been completed.'));
```
Comment (ArchBlood) [inline protected/humhub/components/behaviors/AccessControl.php:251]: Current: `Maintenance mode activated: You have been automatically logged out and will no longer have access the platform until the maintenance has been completed.`

After: `Maintenance mode activated: You have been automatically logged out and will no longer have access to the platform until the maintenance has been completed.`

**Pull request merged**